Changed QuickMeasurement to be conservative

AIRCAP opened the pull request:
fix #16888

QuickMeasurement should not measure while tool dialogs are open this includes but is not limited to editing sketches also changed several other sanity checks to be opt-in vs opt-out, as discussed.

AIRCAP commented on src/Mod/Measure/Gui/QuickMeasure.cpp:
```diff
@@ -111,24 +111,30 @@ void QuickMeasure::tryMeasureSelection()
 
-    std::string vpType = obj->getViewProviderName();
-    auto* vp = Gui::Application::Instance->getViewProvider(obj);
-    return !(vpType == "SketcherGui::ViewProviderSketch" && vp && vp->isEditing());
```
the check for "not editing a sketch" is no longer needed, since editing *anything* will now disable QuickMeasure

PaddleStroke replied:
Minor: In this case you can directly `return obj && obj->isDerivedFrom(Part::Feature::getClassTypeId())`

AIRCAP replied:
IMHO the code is functionally easier to read as is - the compiler will further optimize that when generating machine code anyway

chennes replied:
I'm fine with leaving it like this, as long as the linter doesn't complain about it -- pylint is much crankier about such constructions than our current C++ linters.

hyarion commented:
These changes reflects what was discussed during the developer meeting earlier today:
* summarized by here https://github.com/FreeCAD/FreeCAD/issues/16888#issuecomment-2395508637
* and here https://github.com/FreeCAD/FreeCAD-developer-meetings/blob/main/Minutes/minutes-2023-10-06.md as `reducing traffic to QM / opt-in vs opt-out ("else return true") seems to be a reasonable approach.`

I think this is a good approach as it:
* removes *false positives* that can lead to exceptions and crashes and
* trades that to *potential false negatives*, where the worst thing is that the user don't get a measure.

The outcome of a missing readout might be annoying, but a crash is really bad.

👍 

FYI @PaddleStroke any input on this PR?

AIRCAP commented:
I changed the logic a bit again to be even safer. The issue that the test `shouldMeasure()` was done when the event was triggered - this makes sense as far as the event itself is concerned (is it actually the right type of event) but its dangerous to check the state of the document and Gui there only and then rely on it

because the actual measuring takes place when the timer fires and by then things might have changed.

So in the last commit I moved all tests w.r.t. the state ( particularly document and any Task dialogs open ) to when the timer actually fires. This is when it really matters, and it has the additional benefit that the status line is cleared and no stale measurements remain there while in a Task Dialog.

PaddleStroke commented:
Looks good.

> The outcome of a missing readout might be annoying, but a crash is really bad.

+1.

chennes commented:
Backported to 1.0
